cosa/20200805-meta-merge.md

```diff
@@ -0,0 +1,103 @@
+
+Further, a "failed build" is also used to mean that some part of a pipeline failed to complete. COSA itself does not have any concept of dependencies, much less the idea that a multiple COSA commands can be run. COSA only has implied dependencies that some artifacts requires others. In other words, COSA doesn't know when the "build" is complete until after something else considers its complete. 
+
+Denoting the state of a build in the `meta.json` will be owned by process/thing/person that is orchestrating the build. 
+
+## Mutliple `meta.json`
+
+Under the axiom that the state of a build is usually determined from the point-of-view of a pipeline and to support distributed and parallel builds, builds will be produced with multiple `meta.json` that will be merged together.
+
+## `cosalib.meta` library
+
+`cmd-meta` currently uses the `cosalib.meta` library for reading and writing `meta.json`. To support merging of `meta.json` files, the library will be updated to allow for merging in-memory updates with on-disk content.
+
+Only `meta.json` that is validated will be merged under the following rules:
+- the on-disk version must match in core details for name and buildid for in memory model of the meta
+- on-disk meta.json takes precedence over the in-memory meta
+- all versions must be validated before and after merging before commiting to disk.
+
+Prior to merging `cosalib.meta` will lock the localfile.
+
+## build mode denoted in `meta.json`
+
+Two new build "modes" will be added. `cmd-build` will have build modes written in the `meta.json` of `parallel` (default) or distributed.
+
+```
+{ 
+    "cosa.buildmode": "<parallel|distributed>"
+}
+```
+
+`cmd-build` will have the flag of `--distributed` added to set the build mode.
+
+## Distributed builds
+
+When `cosa.buildmode` is set to `distributed`, `cmd-buildextend.<ARTIFACT>` commands will create suffixed `meta.json.<ARTIFACT>` instead of writing directly to `meta.json`. A new command called `cmd-meta-merge` will be introduced that will read `meta.json.<ARTIFACT>` and will ONLY combine the images and the cloud-build specific sections into `meta.json`. 
+
+Pipelines orchestrating the build will be responsible for shuffling artifacts around. 
+
+All `meta.json.<ARTIFACT>` files will be archived in S3 alongside the `meta.json`. 
+
+`cmd-meta-merge` will be able `meta.json.<ARTIFACT>` located directly in S3. `cmd-meta-merge` will:
+- read in `meta.json` and validate it.
+- iterate over the remaining `meta.json.<ARTIFACT>`. Validation against the schema will happen during the read and after merging.
+
+## Local builds
+
+When `cosa.buildmode` is set to "parallel" `cmd-buildextend-*` the merging of `meta.json` will happen in real time using the `cosalib.meta` Python code to prevent racing conditions. To support bash commands `finalize-artifact` will be replaced with `cmd-meta-merge`.
```

Comment:
```suggestion
When `cosa.buildmode` is set to "parallel", the  `cmd-buildextend-*` commands will manage the merging of `meta.json` will happen in real time using the `cosalib.meta` Python code to prevent racing conditions. To support bash commands `finalize-artifact` will be replaced with `cmd-meta-merge`.
```